runtime/verifier/src/lib.rs

```diff
@@ -34,98 +33,88 @@ impl<'a> TransactionVerifier<'a> {
+            Some(signer) => signer,
+            None => {
+                return Err(format!("Signer {:?} does not exist", signer_id));
+            }
+        };
+        let access_key =
+            match get_access_key(self.state_update, &signer_id, &transaction.public_key) {
+                Some(access_key) => access_key,
+                None => {
                     return Err(format!(
-                        "Transaction nonce {} must be larger than originator's nonce {}",
-                        transaction.nonce, signer.nonce,
+                        "Signer {:?} doesn't have access key with the given public_key {:?}",
+                        signer_id, &transaction.public_key,
                     ));
                 }
+            };
 
-                if !is_valid_account_id(&transaction.receiver_id) {
-                    return Err(format!(
-                        "Invalid receiver account ID {:?} according to requirements",
-                        transaction.receiver_id
-                    ));
-                }
+        if transaction.nonce <= access_key.nonce {
+            return Err(format!(
+                "Transaction nonce {} must be larger than nonce of the used access key {}",
+                transaction.nonce, access_key.nonce,
+            ));
+        }
+
+        if !is_valid_account_id(&transaction.receiver_id) {
+            return Err(format!(
+                "Invalid receiver account ID {:?} according to requirements",
+                transaction.receiver_id
+            ));
+        }
 
-                let hash = signed_transaction.get_hash();
-                let hash = hash.as_ref();
-                let public_key = signer
-                    .public_keys
-                    .iter()
-                    .find(|key| verify(&hash, &signed_transaction.signature, &key))
-                    .cloned();
+        let hash = signed_transaction.get_hash();
+        if !verify(hash.as_ref(), &signed_transaction.signature, &transaction.public_key) {
+            return Err(format!(
+                "Transaction is not signed with a public key of the signer {:?}",
+                signer_id,
+            ));
+        }
 
-                if let Some(public_key) = public_key {
-                    if &public_key != &transaction.public_key {
-                        return Err("Transaction is signed with different public key than given"
-                            .to_string());
+        // TODO: Calculate transaction cost
+
+        match access_key.permission {
+            AccessKeyPermission::FullAccess => Ok(VerificationData {
+                signer_id: signer_id.clone(),
+                signer,
+                public_key: transaction.public_key.clone(),
+                access_key,
+            }),
+            AccessKeyPermission::FunctionCall(ref function_call_permission) => {
+                if transaction.actions.len() != 1 {
+                    return Err(
```

Comment:
Why is this not allowed

Reply:
More than 1 batched function call usually doesn't have a good use-case. Any other non-function-calls actions is denied.